src/Interpreters/TreeRewriter.cpp

```diff
@@ -509,6 +509,186 @@ void setJoinStrictness(ASTSelectQuery & select_query, JoinStrictness join_defaul
     out_table_join = table_join;
 }
 
+class DNF
```

Comment:
Also there's implementation of CNF in this PR, not merged yet

https://github.com/ClickHouse/ClickHouse/pull/18787/files#diff-a7bbea4019513e909dd0d5167dce7160fc5edd6f16a88e71bde00caf9f81703b

How do you thing can we reuse same implementation here?

Reply:
"not" is not covered by DNF builder because it is not allowed as a part of JOIN ON, only ANDs (originally) and ORs (by this PR) are covered.  It might that we should consider adding NOTs into the scope.
It is great about CNF in 18787, well spotted! I think that we basically have to unify these two pieces of code, having them both seems strange. The questions are 'how' and 'when'. I'll look into CNF a bit deeper.

Reply:
I'm not 100% sure that I understand your concern about atomicity and terms, while queries like 'select * from t1 join t2 on t1.a+t2.a=t1.a' or 'select * from t1 join t2 on t1.a+t2.a=t1.a or t1.a=t2.a' produce
```
Code: 403. DB::Exception: Received from localhost:9000. DB::Exception: Invalid columns in JOIN ON section. Columns a and t2.a are from different tables.: While processing (a + t2.a) = a. 
```

Reply:
I meant that `t1.a+t2.a=t1.a` considered as atomic expression and we cannot split it, and the same for `t1.a+t2.a=t1.a or not (t1.a=t2.a)`, so `not (t1.a=t2.a)` cannot be transformed too.